packages/page-experience/lib/PageDataGatherer.js

```diff
@@ -0,0 +1,236 @@
+        }
+        const font = fontFamilyString.split(',')[0];
+        return font.replace(/["']/g, '');
+      };
+
+      /**
+       * Returns a list of critical and non-critical fonts. Critical fonts are used in the first viewport.
+       * All other fonts are considered non-critical.
+       *
+       * TODO: take font-weights into account when calculating critical fonts.
+       *
+       * @return {Object}
+       */
+      const collectFontsUsedOnPage = () => {
+        const criticalFonts = new Set();
+        const nonCriticalFonts = new Set();
+        document.querySelectorAll('body *').forEach((node) => {
+          const computedStyles = window.getComputedStyle(node);
+          const fontFamily = computedStyles.getPropertyValue('font-family');
+          const font = extractFirstFont(fontFamily);
+          if (!font) {
+            return;
+          }
+          if (isElementInViewport(node)) {
+            criticalFonts.add(font);
+          } else {
+            nonCriticalFonts.add(font);
+          }
+        });
+        return {
+          criticalFonts: Array.from(criticalFonts),
+          // Make sure to remove later discovered critical fonts from the list of non-critical ones
+          nonCriticalFonts: Array.from(nonCriticalFonts).filter((font) => !criticalFonts.has(font)),
+        };
+      };
+
+      return {
+        origin: window.location.origin,
+        fontPreloads: collectFontPreloads(),
+        localStyles: collectInlineStyles(),
+        ...collectFontsUsedOnPage(),
+      };
+    });
+
+    return {
+      remoteStyles: remoteStyles,
+      criticalFonts: result.criticalFonts,
+      nonCriticalFonts: result.nonCriticalFonts,
+      fontPreloads: result.fontPreloads,
+      fontFaces: parseFontfaces([...remoteStyles, ...result.localStyles].join('\n'), result.origin),
+    };
+  }
+
+  /**
+   * @private
+   */
+  async setupPage() {
+    const page = await this.browser.newPage();
+    const remoteStyles = [];
+    if (this.debug) {
+      page.on('console', (msg) => console.log('[PAGE LOG] ', msg.text()));
+    }
+    page.setViewport(this.viewport);
+    page.setRequestInterception(true);
+
+    // Abort requests not needed for rendering the page
+    page.on('request', (request) => {
+      const requestTypeIgnoreList = new Set(['image', 'video']);
+      if (requestTypeIgnoreList.has(request.resourceType())) {
+        return request.abort();
+      }
+      if (
+        request.resourceType() === 'script' &&
+        !request.url().startsWith('https://cdn.ampproject.org')
+      ) {
+        // Only donwload AMP runtime scripts as they're need for layouting the page
+        // Once self-hosting is a thing we'll have to change this
+        // TODO: investigate whether we could cache these locally
+        return request.abort();
+      }
+      return request.continue();
+    });
+
+    // Collect external stylesheets from requests as we can't read them otherwise due to CORS
```

Comment:
Is there a way to miss some stuff as a result of this? Any reason not to just disable CORS in rendering options?

Reply:
That didn't work for me.